trace-normalization/src/normalization/normalize_utils.rs

```diff
@@ -0,0 +1,182 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present
+// Datadog, Inc.
+
+use crate::errors;
+
+// MAX_NAME_LEN the maximum length a name can have
+pub const MAX_NAME_LEN: usize = 100;
+// MAX_SERVICE_LEN the maximum length a service can have
+pub const MAX_SERVICE_LEN: usize = 100;
+
+pub const MAX_TAG_LEN: usize = 200;
+
+// TruncateUTF8 truncates the given string to make sure it uses less than limit bytes.
+// If the last character is an utf8 character that would be splitten, it removes it
+// entirely to make sure the resulting string is not broken.
+pub fn truncate_utf8(s: String, limit: usize) -> String {
+    if s.len() <= limit {
+        return s
+    }
+    let mut prev_index = 0;
+    for i in 0..s.len() {
+        if i > limit {
+            return s[0..prev_index].to_string();
+        }
+        prev_index = i;
+    }
+    s
+}
+
+// NormalizeService normalizes a span service and returns an error describing the reason
+// (if any) why the name was modified.
+// TODO: Implement this in a future PR
+// pub fn normalize_service(svc: String, lang: String) -> (String, Option<errors::NormalizeErrors>) {
+    // if svc == "" {
+    //     return (fallback_service(lang), errors::NormalizeErrors::ErrorEmpty);
+    // }
+    // if svc.len() > MAX_SERVICE_LEN {
+    //     return (truncate_utf8(svc, MAX_SERVICE_LEN), errors::NormalizeErrors::ErrorTooLong.into());
+    // }
+    // TODO: implement tag normalization
+    // let s: String = normalize_tag(svc);
+    // if s == "" {
+    //     return (fallbackService(lang), errors::NormalizeErrors::ErrorInvalid)
+    // }
+    // return (s, err)
+    // (svc, None)
+// }
+
+// normalize_name normalizes a span name and returns an error describing the reason
+// (if any) why the name was modified.
+// pub fn normalize_name(name: String) -> (String, Option<errors::NormalizeErrors>) {
+pub fn normalize_name(name: String) -> Result<String, errors::NormalizeErrors> {
```

Comment:
This API means that you will pass a String, and `normalize_name` will take ownership of it, and on error, it will drop it. The name won't exist anymore. This is maybe fine, depends a bit on how you expect it to be used.

There are some other patterns out there, for instance `Cow<str>` might be nice here:

```
pub fn normalize_name(name: &str) -> Result<Cow<str>, errors::NormalizeErrors> {
```

If everything is fine, the string is simply borrowed, never copied. If a change needs to be made, then a new string is created. On an error, the caller still has the original `&str` that was being borrowed.

Reply:
I think` Cow<str>` definitely makes sense here, I'll make this change